examples/java/src/main/java/org/apache/beam/examples/io/examplekafkaread/ReadFromKafka.java

```diff
@@ -0,0 +1,162 @@
+import org.joda.time.Instant;
+
+/**
+ * Because the element here is a topic partition, there can be any number of KV pairs per element.
+ * As such, we mark this as {@link UnboundedPerElement}.
+ *
+ * <p>Note, although base {@link org.apache.beam.sdk.io.kafka.KafkaIO} does provide an @GetSize
+ * method, it is not required for a Kafka based IO to function properly. Because both IOs use an
+ * OffsetRangeTracker, the size of work remaining can be computed using that trackers built-in
+ * getProgress function. In KafkaIO, this is enhanced providing a @GetSize method that improves the
+ * estimate by keeping track of the size of the Kafka elements, but that is not strictly necessary.
+ */
+@UnboundedPerElement
+public class ReadFromKafka extends DoFn<TopicPartition, KV<byte[], byte[]>> {
+  private static final java.time.Duration KAFKA_POLL_TIMEOUT = java.time.Duration.ofSeconds(1);
+
+  final Map<String, Object> consumerConfig;
+
+  public ReadFromKafka(@NonNull Map<String, Object> consumerConfig) {
+    this.consumerConfig = consumerConfig;
+  }
+
+  /**
+   * Our initial restriction from Kafka is the range from the current consumer position until
+   * Long.MAX_VALUE. The reason we do not go from 0 to max value, is that Kafka may have cleaned up
+   * old elements, such that the minimum offset is not 0, or is that our consumer could be using a
+   * group identifier that has already consumed from this topic partition.
+   */
+  @GetInitialRestriction
+  public OffsetRange initialRestriction(@Element TopicPartition topicPartition) {
+    try (Consumer<byte[], byte[]> offsetConsumer =
+        new KafkaConsumer<byte[], byte[]>(consumerConfig)) {
+      offsetConsumer.assign(ImmutableList.of(topicPartition));
+      long startOffset = offsetConsumer.position(topicPartition);
+
+      return new OffsetRange(startOffset, Long.MAX_VALUE);
+    }
+  }
+
+  /**
+   * Because elements can be added to Kafka continuously, we want our offset range tracker to
+   * reflect this. As such, we use a {@link GrowableOffsetRangeTracker} to poll Kafka and see how
+   * far out the end of the range is.
+   *
+   * <p>If the restriction provided by the runner has a getTo() of Long.MAX_VALUE, we don't need to
+   * query Kafka to know how far out to grab our restriction.
+   */
+  @NewTracker
+  public OffsetRangeTracker restrictionTracker(
+      @Element TopicPartition topicPartition, @Restriction OffsetRange restriction) {
+    if (restriction.getTo() < Long.MAX_VALUE) {
+      return new OffsetRangeTracker(restriction);
+    }
+    ExampleKafkaReadOffsetEstimator offsetPoller =
+        new ExampleKafkaReadOffsetEstimator(
+            new KafkaConsumer<byte[], byte[]>(
+                ExampleKafkaReadIOUtils.getOffsetConsumerConfig(
+                    "tracker-" + topicPartition, consumerConfig)),
+            topicPartition);
+    return new GrowableOffsetRangeTracker(restriction.getFrom(), offsetPoller);
```

Comment:
Why is GrowableOffsetRangeTracker useful in this case ? I.e. what would we loose if we just use OffsetRangeTracker(0, Long.MAX_VALUE) for al cases ?